feat(plastic): added Plastic SCM segment

Khaos66 opened the pull request:
### Prerequisites

- [x] I have read and understood the `CONTRIBUTING` guide
- [x] The commit message follows the [conventional commits][cc] guidelines
- [x] Tests for the changes have been added (for bug fixes/features)
- [x] Docs have been added / updated (for bug fixes/features)

### Description

This adds [Plastic SCM](https://www.plasticscm.com/) as a new `plastic` segment. It took inspiration from the `git` segment. 

### Todo

These are points to work on:

- [x] rename `g` to `p`
- [x] remove `StatusColorsEnabled` check
- [x] remove `getCmCommand`
- [x] use `findNamedRegexMatch` more defensively
- [x] refactor `truncateBranch` into new `scm` base type
- [x] refactor `gitStatus` to `scmStatus`
   - [x] switch to new `add` loop logic
- [x] fix `unmerged` counter to reflect files with unhandled merge conflicts
- [x] create new page in `contribute section` with information about setup and manual test of Plastic SCM

[cc]: https://www.conventionalcommits.org/en/v1.0.0/#summary

Khaos66 committed:
feat(plastic): added Plastic SCM segment

Khaos66 commented:
Hi! Please let me know if I missed anything. I'll be happy to contribute =)

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+	FullBranchPath Property = "full_branch_path"
+)
+
+func (g *plastic) enabled() bool {
```
Can you rename `g` to `p` ;-)

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+}
+
+func (g *plastic) string() string {
+	statusColorsEnabled := g.props.getBool(StatusColorsEnabled, false)
```
This is old and needs to be removed, it's only in git for backwards compatibility. We now color the segment based on the color templates.

Khaos66 replied:
Yes, I was wondering what it's used for and forgot to remove it. After all, it's not really used anyway. 
Will be removed!

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+	return text
+}
+
+func (g *plastic) Status() string {
```
This is still based on the previous version of the git segment, it would be interesting to see if the `gitStatus` struct could be reused (and renamed obviously in that case).

Khaos66 replied:
I agree. It could be renamed to `scmStatus`. What do you think?
One difference is that I counted `moved` files separatly and in `git` they are counted as `modified`.

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+	g.parseFilesStatus(splittedOutput)
+}
+
+func (g *plastic) parseFilesStatus(output []string) {
```
Same remark as for `gitStatus`, we can make use (extend) the `add` function there

Khaos66 replied:
I'll have a look at the new `add` logic.
By the way, what does `unmerged` translate to in git? Is it 'files with unhandled merge conflicts' or 'locally merged but uncommited files'? (I implemented the later one for plastic)

JanDeDobbeleer replied:
the first option

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+
+func (g *plastic) parseStatusChangeset(status string) int {
+	var csRegex = `STATUS\s+(?P<cs>[0-9]+?)\s`
+	cs, _ := strconv.Atoi(findNamedRegexMatch(csRegex, status)["cs"])
```
what if there's no match?

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+func (g *plastic) getHeadChangeset() int {
+	output := g.getCmCommandOutput("status", "--head", "--machinereadable")
+	var csRegex = `\bcs:(?P<cs>[0-9]+?)\s`
+	cs, _ := strconv.Atoi(findNamedRegexMatch(csRegex, output)["cs"])
```
same remark as above, no match can crash this

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+	return val
+}
+
+func (g *plastic) getCmCommand() string {
```
This is a duplicate function, if we feel like we need this maybe it's better to add this logic to `runCommand` IF this is an issue for all executables (git is slow in this case so we fallback to the Windows variant)

Khaos66 replied:
Oh. So this might not be needed at all? 
`cm.exe` (windows) and `cm` (others) can both be executed via `cm`. So maybe you'r right. I will test that.

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,245 @@
+	g.env = env
+}
+
+func (g *plastic) truncateBranch(branch string) string {
```
Duplicate function

Khaos66 replied:
True. Where would you refactor this function to? (I need a little guidance here)

JanDeDobbeleer replied:
Good question, we could create an `scm` struct that has the functions (also status) and use struct inheritance in `git` and `plastic`

Khaos66 replied:
This worked out well =) See [edb0fec](https://github.com/JanDeDobbeleer/oh-my-posh/pull/1392/commits/edb0fece458381486862e9e400feb375388231f9)

JanDeDobbeleer reviewed (changes requested):
All in all, really good addition. Some duplications and old style (I admit, I did refactor the git segment like a madman the last few days). About `findNamedRegexMatch`, that's being used a bit too freely and is subject to break if there are no matches.

JanDeDobbeleer committed:
fix(pwsh): alert when we can't download dependencies

resolves #1382

Khaos66 commented:
Wow, thank you for the fast review!
I agree with all you points. I see your changes to `git` in #1389 and will apply them to this segment as well.
It may (or may not) take some time, as I'm doing this in my free time. Cheers

Khaos66 commented:
Also, is there a place where I could document steps for manual testing? I guess it would be helpfull to document the steps to setup a Plastic SCM server and client with repos and workspaces (as this is not common knowlege).
Maybe in a github issue?

Khaos66 committed:
Merge branch 'main' into segment_plastic

JanDeDobbeleer commented:
> Also, is there a place where I could document steps for manual testing? I guess it would be helpfull to document the steps to setup a Plastic SCM server and client with repos and workspaces (as this is not common knowlege).
Maybe in a github issue?

@Khaos66 I would create a separate page in the[ contributing section](https://ohmyposh.dev/docs/contributing_started)

Khaos66 committed:
refactor(plastic): polished new Plastic SCM segment

Khaos66 committed:
refactor(plastic): polished new Plastic SCM segment

Khaos66 committed:
refactor: moved common scm segment code into base type

git and plastic share some common methods and status properties

Khaos66 committed:
refactor: moved common scm segment code into base type

git and plastic share some common methods and status properties. So moving them in a base type keeps the code base DRY

Khaos66 commented:
I'm trying to get the icon for Plastic SCM added to FontAwesome and ultimately to Nerd-Fonts.
Please help me up-vote the issue
https://github.com/FortAwesome/Font-Awesome/issues/18504

Khaos66 committed:
docs(plastic): added docs for manual testing Plastic SCM

Khaos66 committed:
doc(plastic): Added docs for manual testing Plastic SCM

Khaos66 committed:
fix(plastic): Show files with merge conflicts as unmerged

Khaos66 committed:
fix(plastic): show files with merge conflicts as unmerged

Khaos66 commented:
@JanDeDobbeleer I think I'm done. Can you please review again?

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,192 @@
+	selector := p.getFileContents(p.plasticWorkspaceFolder+"/.plastic/", "plastic.selector")
+	// changeset
+	ref = p.parseChangesetSelector(selector)
+	if ref != "" {
```
if `len(ref) == 0` (same as everywhere else)

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,192 @@
+	}
+	// fallback to label
+	ref = p.parseLabelSelector(selector)
+	if ref != "" {
```
if `len(ref) == 0` (same as everywhere else)

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,192 @@
+	}
+	// fallback to branch/smartbranch
+	ref = p.parseBranchSelector(selector)
+	if ref != "" {
```
if `len(ref) == 0` (same as everywhere else)

JanDeDobbeleer commented on src/segment_plastic.go:
```diff
@@ -0,0 +1,192 @@
+
+func (p *plastic) parseIntPattern(output, pattern, name string, defValue int) int {
+	sValue := p.parseStringPattern(output, pattern, name)
+	if sValue != "" {
```
if `len(sValue) == 0` (same as everywhere else)

JanDeDobbeleer commented:
@Khaos66 only 1 styling comment, amazing PR!

Khaos66 committed:
fix(plastic): adhere to empty string check guidelines

Khaos66 committed:
fix(plastic): adhere to empty string check guidelines

Khaos66 commented:
> @Khaos66 only 1 styling comment, amazing PR!

@JanDeDobbeleer Thank you very much! I'm doing what I can =D (first time in `go` though...)

It's done btw ;)

Khaos66 committed:
fix(plastic): fixed linter errors

Khaos66 committed:
fix(plastic): fixed linter errors

Khaos66 committed:
Merge branch 'segment_plastic' of https://github.com/Khaos66/oh-my-posh into segment_plastic

JanDeDobbeleer committed:
Merge branch 'main' into segment_plastic

Khaos66 commented:
<3
